include/nlohmann/detail/output/serializer.hpp

```diff
@@ -801,13 +801,14 @@ class serializer
             }
         };
 
+        assert(0 <= byte and byte < 400);
         const std::uint8_t type = utf8d[byte];
 
         codep = (state != UTF8_ACCEPT)
                 ? (byte & 0x3fu) | (codep << 6u)
                 : (0xFFu >> type) & (byte);
 
-        state = utf8d[256u + state * 16u + type];
+        state = utf8d[256 + (std::uint64_t)state * 16 + (std::uint64_t)type];
```

Comment:
Casting those values to 64 bit unsigned ints to index a 400 element std::array doesn't make much sense for 32 bit platforms and will almost certainly trigger a narrowing warning there. 

Also, why are you turning the integral constants into signed integers (just curious)

Reply:
This is indeed really weird. Could you comment on this?

Reply:
@MBalszun , I will check it for 32 bit paltforms. I just forgot the 32 bit platforms, which i use rarely.

Reply:
I guess the "right" thing would be to cast the values to std::size_t, although I find it somewhat annoying that perfectly fine, safe and readable code needs to be uglified to such a degree (and potentially even become less efficient) just to shut up a compiler warning (or is it a static analysis warning?) .

Reply:
@MBalszun  Had been changed to `std::size_t`

Reply:
Just to be on the safe side, you probably also want the litterals to remain unsigned (i.e. `256u` instead of `256`) It shouldn't make any difference, but I wouldn't be suprised if some compiler/analysis tool would start complaining about the mix of signed and unsigned types.

Reply:
Just to be on the safe side, you probably want to keep the litterals unsigned too (i.e. `256u` instead of `256`) it should not make any difference, but I wouldn't be suprised if some compiler or analysis tool complains about mixing signed and unsigned types.

Reply:
Just to shut up a compiler warning for a single line is not the best way, i think.
In some company, compiler warnings must be solved, so  i understand why people figure these warnings out. They have no choice.

Reply:
Of course the next issue is that a static analysis tool that checks on core-guideline violations will complain about the c-style cast. This is going to be a neverending story. :(
@nlohmann:  Whats this project's policy about c-style casts?